### PR TITLE
swoole custom dispatch function

### PR DESCRIPTION
**Pull request by lohanidamodar**
## What does this PR do?

- Fixes random function timeout

## Test Plan

- manual tests and benchmarks

## Related PRs and Issues

- (Related PR or issue)

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [x] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: app/http.php [hunk 1]
```diff
@@ -327,4 +378,50 @@
     }
 });
 
+function getDomains(Database $dbForConsole, &$lastSyncUpdate, &$domains, $pools)
+{
+    go(function () use ($dbForConsole, &$lastSyncUpdate, &$domains, $pools) {
+        Timer::tick(DOMAIN_SYNC_TIMER * 1000, function () use ($dbForConsole, &$domains, &$lastSyncUpdate, $pools) {
+            $time = DateTime::now();
+            $timerStart = \microtime(true);
+            var_dump($domains);
```
Comment (eldadfux) [inline app/http.php:387]: Reminder

File: app/http.php [hunk 2]
```diff
@@ -36,11 +44,53 @@
...
+function dispatch(Server $server, $fd, $type, $data = null)
+{
+    global $workerNumber, $domains;
+
+    $safeThreadsPercent = intval(App::getEnv('_APP_SAFE_THREADS_PERCENT', 80)) / 100;
+
+    $safeThreads = (int) floor($workerNumber * $safeThreadsPercent);
+
+    $risky = false;
+    if ($data && str_contains($data, 'POST') && str_contains($data, '/executions')) {
```
Comment (Meldiron) [inline app/http.php:62]: $data also includes body, right? I am worried there will be false positive.
Can we split and only get first line for this check? First line should only have method and URL. Then we can also do checks for starts_with and ends_with.

File: app/http.php [hunk 3]
```diff
@@ -36,11 +44,53 @@
...
+        $risky = true;
+    } elseif ($data && str_contains($data, '.appwrite.global')) {
+        $risky = true;
+    } elseif ($data) {
+        $lines = explode($data, '\n');
+        if (count($lines) > 2) {
+            $host = trim(explode('Host: ', $lines[1])[1]);
```
Comment (Meldiron) [inline app/http.php:69]: Hmm, what's most trust-worthy? Host header, or domain parsed from first line (method+url)? Maybe we can use Domains library to simplify the work.

File: app/http.php [hunk 4]
```diff
@@ -36,11 +44,53 @@
...
+            if (array_key_exists($host, $domains)) {
+                $risky = true;
+            }
+        }
+    }
+
+    if ($risky) {
+        var_dump('execution request sending to unsafe thread');
```
Comment (Meldiron) [inline app/http.php:77]: Leftover

File: app/http.php [hunk 5]
```diff
@@ -36,11 +44,53 @@
...
+        for ($j = $safeThreads; $j < $workerNumber; $j++) {
+            if ($server->getWorkerStatus($j) === 2) {
```
Comment (Meldiron) [inline app/http.php:79]: Let's document what 2 means, or use const

File: app/http.php [hunk 6]
```diff
@@ -36,11 +44,53 @@
...
+                return $j;
+            }
+        }
+        return rand($safeThreads, $workerNumber - 1);
+    }
+
+    for ($i = 0; $i < $workerNumber; $i++) {
+        if ($server->getWorkerStatus($i) === 2) {
```
Comment (Meldiron) [inline app/http.php:87]: Let's document what 2 means, or use const if swoole has one.

File: app/http.php [hunk 7]
```diff
@@ -36,11 +44,53 @@
         // 'document_root' => __DIR__.'/../public',
         // 'enable_static_handler' => true,
         'http_compression' => true,
+        'dispatch_func' => 'dispatch',
         'http_compression_level' => 6,
         'package_max_length' => $payloadSize,
         'buffer_output_size' => $payloadSize,
     ]);
 
+function dispatch(Server $server, $fd, $type, $data = null)
```
Comment (eldadfux) [inline app/http.php:53]: Add a docblock to explain our strategy and steps to achieve it.

File: app/http.php [hunk 8]
```diff
@@ -36,11 +44,53 @@
...
+        }
+        return rand($safeThreads, $workerNumber - 1);
+    }
+
+    for ($i = 0; $i < $workerNumber; $i++) {
+        if ($server->getWorkerStatus($i) === 2) {
+            return $i;
+        }
+    }
+    return rand(0, $safeThreads - 1);
```
Comment (Meldiron) [inline app/http.php:91]: Can we do `rand(0, workerNumber - 1)` here instead? If I understand correctly, this is for safe requests. We can use any worker for that.

File: app/http.php (hunk 2 above)
Comment (eldadfux) [inline app/http.php:62]: Wouldn't it be easier to first parse the payload and the set the conditions? It would also be more effective than the extensive usage in str_contains

File: app/http.php (hunk 5 above)
Comment (eldadfux) [inline app/http.php:79]: What is worker status 2? Let's add a reference to relevant docs. Maybe use constants.

File: app/http.php [hunk 9]
```diff
@@ -207,6 +257,8 @@
             $dbForConsole->createCollection('bucket_' . $bucket->getInternalId(), $attributes, $indexes);
         }
 
+        getDomains($dbForConsole, $lastSyncUpdate, $domains, $pools);
+
         $pools->reclaim();
```
Comment (Meldiron) [inline app/http.php:262]: Let's not reclaim here, as dbForConsole is sent to getDomains, and there we have infinite loop.

File: app/http.php [hunk 10]
```diff
@@ -327,4 +378,50 @@
...
+                    }
+                    $domains[$domain] = true;
+                }
+                $latestDocument = !empty(array_key_last($results)) ? $results[array_key_last($results)] : null;
+            }
+
+            $lastSyncUpdate = $time;
+            $timerEnd = \microtime(true);
+
+            $pools->reclaim();
```
Comment (Meldiron) [inline app/http.php:420]: Let's not reclaim. Its fine to steal 1 connection forever, and never put it back to pool.

File: app/http.php [hunk 11]
```diff
@@ -327,4 +378,50 @@
...
+
+            Console::log("Sync domains tick: {$total} domains were updated in " . ($timerEnd - $timerStart) . " seconds");
```
Comment (Meldiron) [inline app/http.php:422]: Hmm, this could end up as spam. Can we print only if there was actually some new rules?

File: app/http.php (hunk 2 above)
Comment (lohanidamodar) [reply app/http.php:62]: We can, but I also worry about the performance?

File: app/http.php [hunk 12]
```diff
@@ -327,4 +378,50 @@
     }
 });
 
+function getDomains(Database $dbForConsole, &$lastSyncUpdate, &$domains, $pools)
+{
+    go(function () use ($dbForConsole, &$lastSyncUpdate, &$domains, $pools) {
```
Comment (Meldiron) [inline app/http.php:383]: Hmm, can we try to add `sleep(10)` in here, all over the place? Just to ensure if anything goes wrong here, master doesn't freeze and can still process HTTP requests.

File: app/http.php (hunk 1 above)
Comment (Meldiron) [inline app/http.php:387]: Leftover

File: app/http.php (hunk 3 above)
Comment (lohanidamodar) [reply app/http.php:69]: First line only has path `/v1` and second line has the host

File: app/http.php [hunk 13]
```diff
@@ -327,4 +378,50 @@
...
+                }
+                $queries[] = Query::equal('resourceType', ['function']);
+
+                $results = Authorization::skip(fn () =>  $dbForConsole->find('rules', $queries));
+
+                $sum = count($results);
+                $total = $total + $sum;
+                foreach ($results as $document) {
+                    $domain = $document->getAttribute('domain');
+                    if (str_contains($domain, '.appwrite.global')) {
```
Comment (Meldiron) [inline app/http.php:409]: str_ends_with is safer

File: app/http.php (hunk 8 above)
Comment (lohanidamodar) [reply app/http.php:91]: no because we reach here only hwne there are no idle threads, and we don't want safe request going to unsafe thread and get blocked in the loop like before.

File: app/http.php (hunk 2 above)
Comment (Meldiron) [reply app/http.php:75]: We can benchmark, split with limit 2 should be very fast - after first split its quick. First split is after method and path, which shouldtn be too much, max ~2k chars.

I would love to avoid this.. Scarry to have a bug that causes random timeouts when there is specific word inside a body of my request

File: app/http.php (hunk 3 above)
Comment (Meldiron) [reply app/http.php:82]: I see.

File: app/http.php (hunk 2 above)
Comment (christyjacob4) [reply app/http.php:70]: I looked into it, splitting by \n itself requires the code to scan through the entire $data.. I don't see a value trying to parse it because the compute we save by removing str_contains is taken up in splitting. Plus, the check the host header requires us to find the host header among all the split lines.

Comment (Meldiron) [reply app/http.php:70]: I dont think full scan is needed if we limit the split,third parameter of https://www.php.net/manual/en/function.explode.php

If we tell it to split only into 2 parts by the first occurrence, I don't want to believe PHP will continue looking for others. We can benchmark it to double check.

File: app/http.php [hunk 14]
```diff
@@ -207,8 +257,6 @@
             $dbForConsole->createCollection('bucket_' . $bucket->getInternalId(), $attributes, $indexes);
         }
 
-        $pools->reclaim();
```
Comment (stnguyen90) [inline app/http.php:210]: Why was this removed?

File: app/http.php [hunk 15]
```diff
@@ -36,11 +40,57 @@
...
+    global $workerNumber;
+
+    $safeThreadsPercent = intval(App::getEnv('_APP_SAFE_THREADS_PERCENT', 80)) / 100;
+
+    $safeThreads = (int) floor($workerNumber * $safeThreadsPercent);
+
+    $risky = false;
+    if ($data && str_contains($data, 'POST') && str_contains($data, '/executions')) {
+        $risky = true;
+    } elseif ($data && str_contains($data, '.appwrite.global')) {
```
Comment (stnguyen90) [inline app/http.php:72]: Can we use the function domain environment variable here instead of hard coding .appwrite.global?

Comment (ekamalov) [reply app/http.php:72]: what happens if we use custom domain?
